### PR TITLE
Fixes #34 Application crash

### LINKED ISSUE #34: Application crash
[AddRef.zip](https://github.com/Open-Systems-Pharmacology/MoBi/files/4230638/AddRef.zip)
Right click on the empty area of the formulas in Events-BB and select `Create Reference...`
MoBi v9 crashes completely without any error message.

In the windows error log: 
<details><summary>Error</summary>

```
Application: MoBi.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
   at OSPSuite.FuncParser.ParsedFunctionImports.SetVariableNames(IntPtr, System.String[], Int32)
   at OSPSuite.FuncParser.ParsedFunction.SetVariableNames(System.Collections.Generic.IEnumerable`1<System.String>)
   at OSPSuite.Core.Services.ExplicitFormulaParser..ctor(System.Collections.Generic.IEnumerable`1<System.String>, System.Collections.Generic.IEnumerable`1<System.String>, OSPSuite.FuncParser.ParsedFunction)
   at OSPSuite.Core.Services.ExplicitFormulaParser..ctor(System.Collections.Generic.IEnumerable`1<System.String>, System.Collections.Generic.IEnumerable`1<System.String>)
   at OSPSuite.Core.Domain.Formulas.FormulaWithFormulaString+<>c.<.cctor>b__15_0(System.Collections.Generic.IEnumerable`1<System.String>, System.Collections.Generic.IEnumerable`1<System.String>)
   at OSPSuite.Core.Domain.Formulas.FormulaWithFormulaString.Validate(System.String)
   at OSPSuite.Core.Domain.Formulas.FormulaWithFormulaString.IsValid(System.String)
   at MoBi.Presentation.Tasks.MoBiFormulaTask.Validate(System.String, OSPSuite.Core.Domain.Formulas.FormulaWithFormulaString, OSPSuite.Core.Domain.Builder.IBuildingBlock)
   at MoBi.Presentation.Presenter.EditExplicitFormulaPresenter.Validate(System.String)
   at MoBi.Presentation.Presenter.EditExplicitFormulaPresenter.Edit(OSPSuite.Core.Domain.Formulas.ExplicitFormula)
   at MoBi.Presentation.Presenter.BasePresenter.AbstractEditPresenter`3[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Edit(System.Object)
   at MoBi.Presentation.Presenter.EditTypedFormulaPresenter`3[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Edit(OSPSuite.Core.Domain.Formulas.IFormula, OSPSuite.Core.Domain.IEntity)
   at MoBi.Presentation.Presenter.EditExplicitFormulaPresenter.Edit(OSPSuite.Core.Domain.Formulas.IFormula, OSPSuite.Core.Domain.IEntity)
   at MoBi.Presentation.Presenter.FormulaCachePresenter.Select(OSPSuite.Core.Domain.Formulas.IFormula)
   at MoBi.Presentation.Presenter.FormulaCachePresenter.Handle(MoBi.Core.Events.FormulaChangedEvent)
   at OSPSuite.Utility.Events.EventPublisher+<>c__DisplayClass4_1`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<PublishEvent>b__1()
   at OSPSuite.Utility.Exceptions.ExceptionManagerBase.Execute(System.Action)
   at OSPSuite.Utility.Events.EventPublisher+<>c__DisplayClass4_1`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<PublishEvent>b__0(System.Object)

Exception Info: System.Reflection.TargetInvocationException
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(System.Object, System.Object[], System.Object[])
   at System.Delegate.DynamicInvokeImpl(System.Object[])
   at System.Windows.Forms.Control.InvokeMarshaledCallbackDo(ThreadMethodEntry)
   at System.Windows.Forms.Control.InvokeMarshaledCallbackHelper(System.Object)
   at System.Windows.Forms.Control.InvokeMarshaledCallback(ThreadMethodEntry)
   at System.Windows.Forms.Control.InvokeMarshaledCallbacks()
   at System.Windows.Forms.Control.MarshaledInvoke(System.Windows.Forms.Control, System.Delegate, System.Object[], Boolean)
   at System.Windows.Forms.Control.Invoke(System.Delegate, System.Object[])
   at System.Windows.Forms.WindowsFormsSynchronizationContext.Send(System.Threading.SendOrPostCallback, System.Object)
   at OSPSuite.Utility.Events.EventPublisher.PublishEvent[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon)
   at MoBi.Core.Domain.Model.MoBiContext.PublishEvent[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon)
   at MoBi.Core.Commands.AddFormulaUsablePathCommand.ExecuteWith(MoBi.Core.Domain.Model.IMoBiContext)
   at OSPSuite.Core.Commands.OSPSuiteCommand`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Execute(System.__Canon)
   at OSPSuite.Core.Commands.Core.CommandExtensions.Run[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.__Canon, System.__Canon)
   at MoBi.Presentation.Tasks.MoBiFormulaTask.AddFormulaUsablePath(OSPSuite.Core.Domain.Formulas.IFormula, OSPSuite.Core.Domain.IFormulaUsablePath, OSPSuite.Core.Domain.Builder.IBuildingBlock)
   at MoBi.Presentation.Presenter.EditFormulaPathListPresenter.AddPathToFormula(OSPSuite.Core.Domain.IFormulaUsablePath)
   at MoBi.Presentation.Presenter.EditFormulaPathListPresenter.CreateNewPath()
   at OSPSuite.Presentation.MenuAndBars.ExecuteActionUICommand.Execute()
   at OSPSuite.Presentation.UICommands.UICommandExtensions.ExecuteWithinExceptionHandler(OSPSuite.Presentation.MenuAndBars.IUICommand, OSPSuite.Utility.Events.IEventPublisher, OSPSuite.Core.Services.IChangePropagator)
   at OSPSuite.Presentation.UICommands.UICommandExtensions.ExecuteWithinExceptionHandler(OSPSuite.Presentation.MenuAndBars.IUICommand)
   at OSPSuite.Presentation.MenuAndBars.MenuBarButton.Click()
   at OSPSuite.UI.Mappers.MenuBarButtonToBarItemMapperBase+<>c__DisplayClass1_0.<UpdateBarButtonItem>b__0(System.Object, DevExpress.XtraBars.ItemClickEventArgs)
   at DevExpress.XtraBars.BarItem.OnClick(DevExpress.XtraBars.BarItemLink)
   at DevExpress.XtraBars.BarButtonItem.OnClick(DevExpress.XtraBars.BarItemLink)
   at DevExpress.XtraBars.BarItemLink.OnLinkClick()
   at DevExpress.XtraBars.BarButtonItemLink.OnLinkAction(DevExpress.XtraBars.BarLinkAction, System.Object)
   at DevExpress.XtraBars.ViewInfo.BarSelectionInfo.ClickLink(DevExpress.XtraBars.BarItemLink)
   at DevExpress.XtraBars.ViewInfo.BarSelectionInfo.UnPressLink(DevExpress.XtraBars.BarItemLink)
   at DevExpress.XtraBars.Controls.CustomLinksControl.OnMouseUp(System.Windows.Forms.MouseEventArgs)
   at System.Windows.Forms.Control.WmMouseUp(System.Windows.Forms.Message ByRef, System.Windows.Forms.MouseButtons, Int32)
   at System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message ByRef)
   at DevExpress.XtraBars.Controls.CustomControl.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr, Int32, IntPtr, IntPtr)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG ByRef)
   at System.Windows.Forms.Application+ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr, Int32, Int32)
   at System.Windows.Forms.Application+ThreadContext.RunMessageLoopInner(Int32, System.Windows.Forms.ApplicationContext)
   at System.Windows.Forms.Application+ThreadContext.RunMessageLoop(Int32, System.Windows.Forms.ApplicationContext)
   at Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase.OnRun()
   at MoBi.UI.Services.MoBiApplication.OnRun()
   at Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase.DoApplicationModel()
   at Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase.Run(System.String[])
   at MoBi.Program.Main(System.String[])
```
</details>

grafik

### PR DESCRIPTION
**Pull request by Yuri05**

### CODE REVIEW

File: src/OSPSuite.FuncParser/ParsedFunction.cs [hunk 1]
```diff
@@ -110,35 +126,36 @@ public bool CaseSensitive
       public void SetVariableNames(IEnumerable<string> variableNames)
       {
          var (variableNamesArray, size) = convertToArray(variableNames);
+         if (variableNamesArray.Any(s => s==null))
+            throw new Exception("Variable name may not be null");
+
```
Comment (Yuri05) [inline src/OSPSuite.FuncParser/ParsedFunction.cs]: Here and in all further cases where STRINGS are passed: checking for string <> null. (that is what caused the crash)
The philosophy of FuncParser is: any invalid inputs are accepted in the first place. All the inputs are evaluated only when `Parse()` or `CalcExpression()` or `GetDimensionInfo()` are called. 
However passing a null strings apparently crashes (e.g. adding null to an std string vector).

Another possibility would be to replace null strings with empty strings in C++ part, but I think passing nulls is absolutely nonsense and should be considered as  an error in the calling app (e.g. in this case in MoBi).

File: src/OSPSuite.FuncParser/ParsedFunction.cs [hunk 2]
```diff
@@ -87,10 +87,26 @@ internal static class ParsedFunctionImports
...
+   {
+      public static void EvaluateCppCallResult(bool success, string errorMessage)
+      {
+         if (success)
+            return;
+
+         throw new Exception(errorMessage);
+      }
+   }
+
```
Comment (Yuri05) [inline src/OSPSuite.FuncParser/ParsedFunction.cs]: just some refactoring

**Comment by msevestre** [comment]
But the code in MoBi has not changed. So this was working ok with null
before and not anymore? Suddenly we have to change the client of the lib...
Why not support this use case as we had it previously?

**Comment by Yuri05** [comment]
you are right; I've created the new PR.
Looking at the old implementation in C++/CLI: [here](https://github.com/Open-Systems-Pharmacology/OSPSuite.FuncParser/blob/f6d2307af2102100a7bcb1f9b526a08963a99d8e/src/OSPSuite.FuncParser/managed/src/Conversions.cpp#L14) null strings were replaced by empty strings.
In the current PInvoke implementation such a conversion was not required (string arrays passed from C# can be directly interpreted as const char* array) - that's why it crashed later.

**Pull request merged**